test/unit/plugin/v2/installer_test.rb

```diff
@@ -0,0 +1,414 @@
+
+    # Verify presence of gemspecs
+    spec_path = File.join(@installer.gem_path, 'specifications', 'inspec-test-fixture-0.2.0.gemspec')
+    assert File.exist?(spec_path), 'After update, the 0.2.0 gemspec should be installed to the gem path'
+    spec_path = File.join(@installer.gem_path, 'specifications', 'inspec-test-fixture-0.1.0.gemspec')
+    assert File.exist?(spec_path), 'After update, the 0.1.0 gemspec should remain'
+
+    # Plugins file entry should be version pinned
+    plugin_json_path = File.join(ENV['INSPEC_CONFIG_DIR'], 'plugins.json')
+    plugin_json_data = JSON.parse(File.read(plugin_json_path))
+    entry = plugin_json_data['plugins'].detect { |e| e["name"] == 'inspec-test-fixture'}
+    assert_includes entry.keys, 'version', 'plugins.json should include version pinning key'
+    assert_equal '= 0.2.0', entry['version'], 'plugins.json should include version pinning value'
+  end
+
+  # TODO: Prevent updating a gem if it will lead to unsolveable dependencies
+  # TODO: allow updating a gem that will lead to unsolveable dependencies if :force is provided
+  # TODO: Prevent downgrading a gem if it will lead to unsolveable dependencies
+  # TODO: allow downgrading a gem that will lead to unsolveable dependencies if :force is provided
+  # TODO: update all
+  # TODO: downgrade a plugin
+  # TODO: Trying to do a gemfile install with an update is an error if the file version matches the installed version
+
+end
+
+#-----------------------------------------------------------------------#
+# Uninstalling
+#-----------------------------------------------------------------------#
+class PluginInstallerUninstallTests < MiniTest::Test
+  include InstallerTestHelpers
+
+  def test_uninstalling_a_nonexistant_plugin_is_an_error
+    # Try a mythical one
+    ex = assert_raises(Inspec::Plugin::V2::UnInstallError) do
+      @installer.uninstall('inspec-test-fixture-nonesuch')
+    end
+    assert_includes ex.message, "'inspec-test-fixture-nonesuch' is not installed, refusing to uninstall."
+
+    # Try a real plugin that is not installed
+    ex = assert_raises(Inspec::Plugin::V2::UnInstallError) do
+      @installer.uninstall('inspec-test-fixture')
+    end
+    assert_includes ex.message, "'inspec-test-fixture' is not installed, refusing to uninstall."
+  end
+
+  def test_uninstalling_a_path_based_plugin_works
+    copy_in_config_dir('meaning_by_path')
+    ENV['INSPEC_CONFIG_DIR'] = File.join(@config_dir_path, 'empty')
+    @installer.__reset_loader
+
+    @installer.uninstall('inspec-meaning-of-life')
+
+    # Plugins file entry should be removed
+    plugin_json_path = File.join(ENV['INSPEC_CONFIG_DIR'], 'plugins.json')
+    plugin_json_data = JSON.parse(File.read(plugin_json_path))
+    entries = plugin_json_data['plugins'].select { |e| e["name"] == 'inspec-meaning-of-life'}
+    assert_empty entries, "After path-based uninstall, plugin name should be removed from plugins.json"
```

Comment:
Should we also check that the gem was removed from the universe for these remove cases?

Reply:
For path, no change: the gem system is unaware of the path-based install.

For the others.... great idea!